// @tarojs/components/dist/esm-es5/taro-canvas-core.entry.js
import { r as registerInstance, c as createEvent, h, g as getElement } from './core-42efd82a.js';
var LONG_TAP_DELAY = 500;
var Canvas = /** @class */ (function () {
    function Canvas(hostRef) {
        var _this = this;
        registerInstance(this, hostRef);
        this.width = 300;
        this.height = 150;
        this.onTouchStart = function () {
            _this.timer = setTimeout(function () {
                _this.onLongTap.emit();
            }, LONG_TAP_DELAY);
        };
        this.onTouchMove = function () {
            clearTimeout(_this.timer);
        };
        this.onTouchEnd = function () {
            clearTimeout(_this.timer);
        };
        this.onLongTap = createEvent(this, "longtap", 7);
    }
    Canvas.prototype.componentDidLoad = function () {
        var _a = this.el.getBoundingClientRect(), width = _a.width, height = _a.height;
        this.width = width;
        this.height = height;
        this.klass = this.el.className;
        this.css = this.el.style.cssText;
    };
    Canvas.prototype.componentDidUpdate = function () {
        var _a = this.el.getBoundingClientRect(), width = _a.width, height = _a.height;
        if (this.width !== width)
            this.width = width;
        if (this.height !== height)
            this.height = height;
        if (this.canvas) {
            this.canvas.className = this.el.className;
            this.canvas.style.cssText = this.el.style.cssText;
        }
    };
    Canvas.prototype.render = function () {
        var _this = this;
        var _a = this, canvasId = _a.canvasId, width = _a.width, height = _a.height;
        return (h("canvas", { "canvas-id": canvasId, ref: function (node) { return (_this.canvas = node); }, width: width, height: height, onTouchStart: this.onTouchStart, onTouchMove: this.onTouchMove, onTouchEnd: this.onTouchEnd }));
    };
    Object.defineProperty(Canvas.prototype, "el", {
        get: function () { return getElement(this); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Canvas, "style", {
        get: function () { return "taro-canvas-core{position:relative;display:block;width:300px;height:150px}"; },
        enumerable: true,
        configurable: true
    });
    return Canvas;
}());
export { Canvas as taro_canvas_core };
